plugins/notifications-backend/src/service/auth.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { AuthenticationError, NotAllowedError } from '@backstage/errors';
+import { getBearerTokenFromAuthorizationHeader } from '@backstage/plugin-auth-node';
+import {
+  AuthorizeResult,
+  BasicPermission,
+  PermissionEvaluator,
+} from '@backstage/plugin-permission-common';
+
+import express from 'express';
+
+import { DefaultServiceUser } from './constants';
+import { RouterOptions } from './types';
+
+export type GetLoggedInUserOptions = Pick<
+  RouterOptions,
+  'identity' | 'tokenManager' | 'externalCallerSecret'
+>;
+export type CheckUserPermission = GetLoggedInUserOptions &
+  Pick<RouterOptions, 'permissions'>;
+
+/*
+ * User's entity must be present in the catalog.
+ */
+export const getLoggedInUser = async (
+  request: express.Request,
+  { identity, tokenManager, externalCallerSecret }: GetLoggedInUserOptions,
+): Promise<string> => {
+  const identityResponse = await identity.getIdentity({ request });
+
+  // To properly set identity, see packages/backend/src/plugins/auth.ts or https://backstage.io/docs/auth/identity-resolver
+  if (identityResponse) {
+    // The auth token contains user's identity, most probably originated in the FE
+    // Example: user:default/guest
+    let author = identityResponse?.identity.userEntityRef;
+    if (author) {
+      if (author.startsWith('user:')) {
+        author = author.slice('user:'.length);
```

Comment:
Why would you remove the `user:` from the entityRef?

Reply:
We could keep it but then we would need to either require or adjust `targetUsers` and `targetGroups` for the same format. It seems to be not friendly there but it is a "hidden detail" here.
Do you think there might be a collision?

Reply:
Or you could simplify this so that it can target only users - see my other comment about the groups

Reply:
the prefix 'kind:' is needed only when a single field/string is used for all kinds. we have a separate table for each type. therefore we do not need the kind string.

Reply:
Full entity refs (kind and all) are extremely commonplace throughout backstage - they are the chosen canonical identifier used both within and across systems in general and something that devs grow to expect to see.

This is not a review or opinion yet on the exact code here (I'll be back from vacation on Monday), but I thought it might give some flavor here.

Reply:
saying 'others do it' is not a valid reason to do something. Especially when others might be wrong (first was wrong and all the rest copy pasted). Provide a real scenario where it is needed and we'll change

Reply:
I kinda designed that system though :)

Look, the idea is that [string refs](https://backstage.io/docs/features/software-catalog/references#string-references) are atomic unique identifiers for entities by their very design. And they are effectively opaque; the fact that they happen to contain the verbatim kind+namespace+name is "accidental" (sometimes I wish I had just base64 encoded them or something) and not meant to be parsed or inspected. It could literally just as well have been a UUID; we chose this specific format only because it was helpful for human readers to make sense of when debugging etc.

The refs carry subtle semantics like that they are lowercased and in what exact locale we do that, and what happens if one of the parts of the id contains colons and slashes, etc. We should basically even be able to redesign how these IDs are made and no code should break. Yeah in reality we won't since these identifiers are persisted and uses as primary keys in so many places, but still.

Consider also that this entity model is extremely open to extension; the User and Group kinds are only suggestions and others may extend or omit or tweak things in their private implementations. Ideally, code should be as "kind-agnostic" as it can be, which of course only works up to a certain point.

This is all slightly academic, one can argue. But I hope that this background was helpful in illustrating the mindset. There are still places where people choose to parse the refs against better judgment and I'm not sure I want to strictly stop you from doing so. But a future reader of the code might very well think "wtf why are these refs partial, whyyy, now i have to compensate for that"